### PR TITLE
Migrate to esConsumes HIP and MillePede Alignment Algorithms

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

Part of #31061.
Its purpose is to migrate the core base classes of `Alignment/CommonAlignmentAlgorithm` and percolate it down to `MillePedeAlignmentAlgorithm` and `HIPAlignmentAlgorithm`.

#### PR validation:

It compiles and runs successfully wf. 1001.0 (it exercises the MillePede algo via the `SiPixelAli` PCL worfklow).
In addition run a SiPixelAli workflow with more statistics:
```
cmsDriver.py milleStep --data --conditions auto:run3_data_express --triggerResultsProcess RECO --scenario pp --datatier ALCARECO --eventcontent ALCARECO --era Run2_2018 -s ALCA:PromptCalibProdSiPixelAli --dasquery='file dataset=/StreamExpress/Run2018B-TkAlMinBias-Express-v1/ALCARECO run=317320' -n 10000 --processName=ReAlCA
```
followed by:
```
cmsDriver.py pedeStep --data --conditions auto:run3_data_express --era Run2_2018 --scenario pp -s ALCAHARVEST:SiPixelAli --filein file:PromptCalibProdSiPixelAli.root 
```
A more thorough validation from the Tracker Alignment group is available [here](https://indico.cern.ch/event/1000811/contributions/4409493/attachments/2265304/3846194/meeting_16_06.pdf).

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport, no backport is needed.
cc:
@antoniovagnerini @vbotta

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
@cmsbuild, please test

File: Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc [hunk 1]
```diff
@@ -49,7 +47,29 @@ AlignmentProducerBase::AlignmentProducerBase(const edm::ParameterSet& config)
...
       useSurvey_{config.getParameter<bool>("useSurvey")},
       enableAlignableUpdates_{config.getParameter<bool>("enableAlignableUpdates")},
-      idealGeometryLabel("idealForAlignmentProducerBase") {
+      idealGeometryLabel("idealForAlignmentProducerBase"),
+      ttopoToken_(iC.esConsumes<edm::Transition::BeginRun>()),
+      geomDetToken_(iC.esConsumes<edm::Transition::BeginRun>()),
+      ptpToken_(iC.esConsumes<edm::Transition::BeginRun>()),
+      //dtGeomToken_(iC.esConsumes<edm::Transition::BeginRun>(edm::ESInputTag("", "idealForAlignmentProducerBase"))),
+      //cscGeomToken_(iC.esConsumes<edm::Transition::BeginRun>(edm::ESInputTag("", "idealForAlignmentProducerBase"))),
+      //gemGeomToken_(iC.esConsumes<edm::Transition::BeginRun>(edm::ESInputTag("", "idealForAlignmentProducerBase"))),
```
Comment (mmusich) [inline Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: @cvuosalo @slomeo 
do you have any hunch as to why uncommenting these lines results in this error:
```
----- Begin Fatal Exception 03-Jun-2021 14:14:58 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Running EventSetup component MuonGeometryConstantsESModule/'muonGeometryConstants
Exception Message:
No data of type "DDCompactView" with label "" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```
despite having the Muon geometry `ESProducers` included here:
https://github.com/cms-sw/cmssw/blob/a7f2449ccf521bd56aea903371f05064a0f23b6a/Alignment/CommonAlignmentProducer/python/AlignmentProducerAsAnalyzer_cff.py#L79-L103

?

Comment (cvuosalo) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: The `DDCompactViewESProducer` needs to be added:
```
process.DDCompactViewESProducer = cms.ESProducer("DDCompactViewESProducer",
                                                  appendToDataLabel = cms.string('')
)
```
Also, change all instances of `fromDDD` to False and `fromDD4hep` to True.

Comment (mmusich) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: @cvuosalo 
this results in:
```
----- Begin Fatal Exception 04-Jun-2021 09:25:50 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Running EventSetup component DDCompactViewESProducer/'DDCompactViewAlignmentProducerAsAnalyzer
Exception Message:
No data of type "DDDetector" with label "" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------

```

Comment (cvuosalo) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: It is probably best to import `Configuration/​Geometry/​python/​GeometryDD4hep_cff.py` to get the necessary producers.

Comment (mmusich) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: @cvuosalo 
adding:
```diff
+from Configuration.Geometry.GeometryDD4hep_cff import *
```
results in:

```
---- Begin Fatal Exception 07-Jun-2021 13:04:47 CEST-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named the.py
Exception Message:
 unknown python problem occurred.
RuntimeError: An exception of category 'FileInPathError' occurred.
Exception Message:
Relative path must not be empty


At:
  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-03-1100/python/FWCore/ParameterSet/Types.py(835): insertInto
  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-03-1100/python/FWCore/ParameterSet/Mixins.py(378): insertContentsInto
  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-03-1100/python/FWCore/ParameterSet/Mixins.py(513): insertInto
  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-03-1100/python/FWCore/ParameterSet/Config.py(1104): _insertManyInto
  /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-03-1100/python/FWCore/ParameterSet/Config.py(1337): fillProcessDesc
  <string>(2): <module>

----- end Fatal Exception -------------------------------------------------
```
and if I add:

```
+DDDetectorESProducer.confGeomXMLFiles = cms.FileInPath("Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml")
```

I go back to initial error:
```
----- Begin Fatal Exception 07-Jun-2021 17:22:39 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Running EventSetup component MuonGeometryConstantsESModule/'muonGeometryConstants
Exception Message:
No data of type "DDCompactView" with label "" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```

Comment (cvuosalo) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: You might try adding:
```
 process.DDDetectorESProducer = cms.ESSource("DDDetectorESProducer",
                                             confGeomXMLFiles = cms.FileInPath('Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml'),
                                             appendToDataLabel = cms.string('')
                                             )
process.DDCompactViewESProducer = cms.ESProducer("DDCompactViewESProducer",
                                                  appendToDataLabel = cms.string('')
                                        )
```

Comment (mmusich) [reply Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:56]: unfortunately none of the suggestions seems to work. As I don't have time to investigate further and the Tracker part is working as expected, I am opening the PR for review and I'll let he muon experts to tackle that.

**Comment by mmusich** [comment]
@cmsbuild, please test

to refresh the tests.

**Comment by yuanchao** [comment]
+1

* migrate to esComsumes (incomplete)
* Alignment/CommonAlignmentAlgorithm to MillePedeAlignmentAlgorithm and HIPAlignmentAlgorithm
* wf matrix tests and unit tests passed

**Comment by qliphy** [comment]
+1

**Comment by smuzaffar** [comment]
@mmusich , there is unit test failing in last night IB [a]. I do not know ifchange in this PR has caused that test to fail, can you please check? Thanks

[a] https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-06-22-2300/unitTestLogs/Alignment/TrackerAlignment#/12-12
```
' looking for alignment labels ''.
Finds tag HCAL : 1
%MSG-i HCAL:  (NoModuleName) 23-Jun-2021 08:40:36 CEST pre-events
CaloTowerTopologyEP::CaloTowerTopologyEP
%MSG
' looking for alignment labels ''.
%MSG-i Alignments:  (NoModuleName) FakeAlignmentProducer()  23-Jun-2021 08:40:37 CEST pre-events
Providing data with label 'fakeForIdeal'.
%MSG
GEMGeometryESModule::initailized with flags 1:0
%MSG-i SiPixelQualityESProducer::SiPixelQualityESProducer:  (NoModuleName)  23-Jun-2021 08:40:37 CEST pre-events

%MSG
%MSG-i SiStripBackPlaneCorrectionDepESProducer:  (NoModuleName)  23-Jun-2021 08:40:37 CEST pre-events
ctor
%MSG
%MSG-i SiStripLorentzAngleDepESProducer:  (NoModuleName)  23-Jun-2021 08:40:37 CEST pre-events
ctor
%MSG
%MSG-i SiStripQualityESProducer:  (NoModuleName) 23-Jun-2021 08:40:37 CEST  pre-events
ctor
%MSG
%MSG-i Geometry:  (NoModuleName) TrackerDigiGeometryESModule()  23-Jun-2021 08:40:37 CEST pre-events
Label 'idealForDigi' IGNORING alignment labels 'fakeForIdeal'.
%MSG
%MSG-i TRACKER:  (NoModuleName) 23-Jun-2021 08:40:38 CEST pre-events
TrackerTopologyEP::TrackerTopologyEP
%MSG
----- Begin Fatal Exception 23-Jun-2021 08:40:41 CEST-----------------------
An exception of category 'LogicError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing Looper: class=AlignmentProducer label='@main_looper'
Exception Message:
A module declared it consumes an EventSetup product after its constructor.
This must be done in the contructor
The product type was: TrackerTopology in record TrackerTopologyRcd
and ESInputTag was Module label:  Data label:
----- End Fatal Exception -------------------------------------------------
```

**Comment by mmusich** [comment]
@smuzaffar it it likely that this PR caused the problem in the unit test.
Out of curiosity from where do you read the log you pasted? It's not displayed in the IB webpage

**Comment by smuzaffar** [comment]
@mmusich , I ran the test manually. May be you can update the test to dump the log files at

https://github.com/cms-sw/cmssw/blob/master/Alignment/TrackerAlignment/test/pixelPositions.sh#L110
https://github.com/cms-sw/cmssw/blob/master/Alignment/TrackerAlignment/test/pixelPositions.sh#L146

so that one can see the contents in case of failures

**Comment by mmusich** [comment]
unfortunately I don't see immediately what's the problem (and even how to start debugging it). 
Perphaps @makortel has a good suggestion.

**Comment by makortel** [comment]
I was able to reproduce, and am taking a look.

**Comment by makortel** [comment]
The immediate problem is the order of base classes for the multiple inheritance of `AlignmentProducer`
https://github.com/cms-sw/cmssw/blob/5aa4fdee130a66887ba8195fa6904eb977f60d72/Alignment/CommonAlignmentProducer/plugins/AlignmentProducer.h#L15
The `consumesCollector()` is a function via the second, `edm::ESProducerLooper()`, and the `edm::ConsumesCollector` it returns is given to the constructor of the first, `AlignmentProducerBase`
https://github.com/cms-sw/cmssw/blob/5aa4fdee130a66887ba8195fa6904eb977f60d72/Alignment/CommonAlignmentProducer/plugins/AlignmentProducer.cc#L13-L14
The `ConsumesCollector` holds a pointer to a base class of `ESProducerLooper`, `edm::EDConsumerBase`, that checks if the consumes registration call is in the constructor. But now the `EDConsumerBase` has not been constructed yet, we get undefined behavior that shows up as this exception being shown.

The fix would be to change the order of the base classes. I tried that, but then I get
```
----- Begin Fatal Exception 23-Jun-2021 16:20:44 CEST-----------------------
An exception of category 'MakeDataException' occurred while
   [0] Using EventSetup component TrackerTopologyEP/'trackerTopology' to make data TrackerTopology/'' in record TrackerTopologyRcd
Exception Message:
Error while making data "TrackerTopology" "" in Record TrackerTopologyRcd
----- End Fatal Exception -------------------------------------------------
```
followed by a segfault
```
#3  0x00007efed087459b in (anonymous namespace)::sig_dostack_then_abort (sig=<optimized out>) at .../CMSSW_12_0_X_2021-06-22-2300/src/FWCore/Services/plugins/InitRootHandlers.cc:539
#4  <signal handler called>
#5  0x00007efea9c21e84 in AlignmentParameterStore::restoreCachedTransformations(unsigned int const&) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-22-2300/lib/slc7_amd64_gcc900/libAlignmentCommonAlignmentAlgorithm.so
#6  0x00007efea95e1cf5 in MillePedeAlignmentAlgorithm::setParametersForRunRange(std::pair<unsigned int, unsigned int> const&) () from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-22-2300/lib/slc7_amd64_gcc900/pluginAlignmentMillePedeAlignmentAlgorithmAuto.so
#7  0x00007efeaa09e6f7 in AlignmentProducerBase::storeAlignmentsToDB (this=0x7efeaccd72c0) at /cvmfs/cms-ib.cern.ch/nweek-02686/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/unique_ptr.h:360
#8  0x00007efeaa09eef6 in AlignmentProducerBase::finish (this=this@entry=0x7efeaccd72c0) at .../CMSSW_12_0_X_2021-06-22-2300/src/Alignment/CommonAlignmentProducer/src/AlignmentProducerBase.cc:784
#9  0x00007efeaa110577 in AlignmentProducer::endOfJob (this=0x7efeaccd6e10) at .../CMSSW_12_0_X_2021-06-22-2300/src/Alignment/CommonAlignmentProducer/plugins/AlignmentProducer.cc:40
#10 0x00007efed8cb9bda in std::function<void ()>::operator()() const (this=this@entry=0x7ffefae2f290) at /cvmfs/cms-ib.cern.ch/nweek-02686/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/std_function.h:683
#11 edm::ExceptionCollector::<lambda()>::operator() (__closure=<synthetic pointer>) at .../CMSSW_12_0_X_2021-06-22-2300/src/FWCore/Utilities/src/ExceptionCollector.cc:40
#12 edm::convertException::wrap<edm::ExceptionCollector::call(std::function<void()>)::<lambda()> > (iFunc=...) at .../CMSSW_12_0_X_2021-06-22-2300/src/FWCore/Utilities/interface/ConvertException.h:21
#13 edm::ExceptionCollector::call(std::function<void ()>) (this=this@entry=0x7ffefae2f2b0, f=...) at .../CMSSW_12_0_X_2021-06-22-2300/src/FWCore/Utilities/src/ExceptionCollector.cc:40
#14 0x00007efed905159c in edm::EventProcessor::endJob (this=0x7efed2163300) at /cvmfs/cms-ib.cern.ch/nweek-02686/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/std_function.h:87
#15 0x000000000040b83c in (anonymous namespace)::EventProcessorWithSentry::~EventProcessorWithSentry() ()
#16 0x000000000040b649 in main ()
```

**Comment by makortel** [comment]
I opened an issue #34223 to follow up

**Pull request merged**